File: src/main/java/Solution_206_ReverseLinkedList.java
public class Solution_206_ReverseLinkedList {
    public static void main(String[] args) {
        var node1 = new ListNode2(1);
        var node2 = new ListNode2(2);
        var node3 = new ListNode2(3);
        var node4 = new ListNode2(4);
        var node5 = new ListNode2(5);
        node1.setNext(node2);
        node2.setNext(node3);
        node3.setNext(node4);
        node4.setNext(node5);
        System.out.println(reverseListIteratively(node1));
        System.out.println(reverseListRecursively(node1));
    }

    private static ListNode2 reverseListIteratively(ListNode2 head) {
        ListNode2 prev = null;
        var current = head;

        while (current != null) {
            var nextTemp = current.next;
            current.next = prev;
            prev = current;
            current = nextTemp;
        }

        return prev;
    }

    private static ListNode2 reverseListRecursively(ListNode2 head) {
        if (head == null || head.next == null) return head;

        var reversedList = reverseListRecursively(head.next);
        head.next.next = head;
        head.next = null;

        return reversedList;
    }
}

class ListNode2 {
    final int val;
    ListNode2 next = null;

    public ListNode2(int val) {
        this.val = val;
    }

    public int getVal() {
        return val;
    }

    public ListNode2 getNext() {
        return next;
    }

    public void setNext(ListNode2 next) {
        this.next = next;
    }
}

File: src/main/kotlin/Solution_206_ReverseLinkedList.kt
fun main() {
    val node1 = ListNode(1)
    val node2 = ListNode(2)
    val node3 = ListNode(3)
    val node4 = ListNode(4)
    val node5 = ListNode(5)
    node1.next = node2
    node2.next = node3
    node3.next = node4
    node4.next = node5
    println(reverseListIteratively(node1))
    println(reverseListRecursively(node1))
}

/**
 * iteratively
 */
fun reverseListIteratively(head: ListNode?): ListNode? {
    var prev: ListNode? = null
    var current = head
    while (current != null) {
        val nextTemp = current.next
        current.next = prev
        prev = current
        current = nextTemp
    }
    return prev
}

/**
 * Recursively
 *
 * Рекурсия идет до конца списка, пока не достигнет последнего узла (это будет новый головной узел после разворота).
 *
 * На каждом шаге возврата из рекурсии:
 *
 * Текущий узел head становится следующим для своего бывшего следующего узла (head.next.next = head).
 *
 * Ссылка next текущего узла устанавливается в null (чтобы избежать циклов).
 *
 * В конце возвращается новый головной узел (бывший последний узел исходного списка).
 *
 * Пример работы:
 * Исходный список: 1 → 2 → 3 → 4 → 5
 * Развернутый список: 5 → 4 → 3 → 2 → 1
 *
 * Сложность этого решения: O(n) по времени и O(n) по памяти (из-за стека рекурсивных вызовов).
 */
fun reverseListRecursively(head: ListNode?): ListNode? {
    // Базовый случай рекурсии: если список пуст или содержит один элемент
    if (head?.next == null) return head

    // Рекурсивно разворачиваем остаток списка
    val reversedList = reverseListRecursively(head.next)

    // Переворачиваем ссылки
    head.next?.next = head
    head.next = null

    // Возвращаем новый головной узел
    return reversedList
}


class ListNode(val value: Int) {
    var next: ListNode? = null
}